File: ValidSudoku.c
//The code is fine but come under runtime error in leetcode, I think some space error
//The python3 version of the same code works properly in this case.
#include<stdio.h>
#include<stdlib.h>
#include<string.h>
#include<stdbool.h>


bool isValidSudoku(char** board, int boardSize, int *boardColSize) {
    int k=0;
    bool valid = true;
    while(k<9){
        int charcount[256] = {0};
        int charcount1[256] = {0};
        for(int j=0;j<9;j++){
            if(board[k][j] != '.'){
                charcount[(int)board[k][j]]++;
                if(charcount[(int)board[k][j]] > 1){
                    valid = false;
                }
            } 
            if(board[j][k]!='.'){
                charcount1[(int)board[j][k]]++;
                if(charcount1[(int)board[j][k]]>1){
                    valid = false;
                }
            }
        }
        k++;
    }
    int i=1,j=1;
    while(i<8){
        while(j<8){
            int charcount3[256] = {0};
            for(int l=i-1;l<i+2;l++){
                for(int h=j-1;j<j+2;h++){
                    if(board[l][h] != '.'){
                        charcount3[(int)board[l][h]]++;
                        if(charcount3[(int)board[l][h]]>1){
                            valid = false;
                        }
                    }
                }
            }
            j+=3;
        }
        i+=3;
    }
    return valid;
}
int main(){
    char sudoku[10][10];
    /*for(int i=0;i<9;i++){
        sudoku[i]=(char)malloc(9*sizeof(char));
    }*/
    for(int i=0;i<3;i++){
        for(int j=0;j<3;j++){
            scanf("%c\n", &sudoku[i][j]);
        }
    }
    bool isit = isValidSudoku(sudoku, 3, 3);
    if(isit == true){
        printf("true\n");
    }
    else{
        printf("false\n");
    }
}

File: ValidSudoku.py
#Only the solution is written here not the whole program.
#Time Complexity = O(1)
#Space Co,plexity = O(1)
#Values given by leetcode.
class Solution:
    def isValidSudoku(self, board: list[list[str]]) -> bool:
        valid = True
    
    # Check rows and columns
        for k in range(9):
            charcount = [0] * 256  # For rows
            charcount1 = [0] * 256  # For columns
            
            for j in range(9):
                if board[k][j] != '.':
                    charcount[ord(board[k][j])] += 1
                    if charcount[ord(board[k][j])] > 1:
                        valid = False
                
                if board[j][k] != '.':
                    charcount1[ord(board[j][k])] += 1
                    if charcount1[ord(board[j][k])] > 1:
                        valid = False

        # Check 3x3 subgrids
        i = 1
        while i < 8:
            j = 1
            while j < 8:
                charcount3 = [0] * 256
                for l in range(i - 1, i + 2):
                    for h in range(j - 1, j + 2):
                        if board[l][h] != '.':
                            charcount3[ord(board[l][h])] += 1
                            if charcount3[ord(board[l][h])] > 1:
                                valid = False
                j += 3
            i += 3
        
        return valid
            